runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexParDoOperator.java

```diff
@@ -393,4 +506,71 @@ public void clear() {
 
   }
 
+  private class CurrentKeyTimerInternals<K> implements TimerInternals {
+
+    private TimerInternalsFactory<K> factory = new TimerInternalsFactory<K>() {
+      @Override
+      public TimerInternals timerInternalsForKey(K key) {
+        InMemoryTimerInternals timerInternals = perKeyTimerInternals.get(key);
+        if (timerInternals == null) {
+          System.out.println("new timer internals for " + key);
+          perKeyTimerInternals.put(key, timerInternals = new InMemoryTimerInternals());
+        }
+        return timerInternals;
+      }
+    };
+
+    // TODO: fault tolerant timers
+    final Map<K, InMemoryTimerInternals> perKeyTimerInternals = new HashMap<>();
```

Comment:
In the case of SDF I believe that is the case, in general not. I think this needs to be addressed before we can say that the runner supports timers.